pkg/controller/utils/datadog/datadog.go

```diff
@@ -0,0 +1,441 @@
+	if err != nil {
+		log.Error(err, "cannot get Datadog credentials")
+		return err
+	}
+	if err := dd.updateCredsIfNeeded(apiKey, appKey); err != nil {
+		log.Error(err, "cannot update Datadog credentials")
+		return err
+	}
+	log.Info("Collecting metrics and events")
+	dd.setTags(dad)
+	status := dad.Status.DeepCopy()
+	if err := dd.sendStatusMetrics(status); err != nil {
+		log.Error(err, "cannot send status metrics to Datadog")
+		return err
+	}
+	if err := dd.sendStatusEvents(status); err != nil {
+		log.Error(err, "cannot send events to Datadog")
+		return err
+	}
+	return nil
+}
+
+// sendStartupEvent posts the operator startup event
+// it must be called only once, when the operator starts
+func (dd *MetricsForwarder) sendStartupEvent() error {
+	event := &api.Event{
+		Time:      api.Int(int(time.Now().Unix())),
+		Title:     api.String("Datadog Operator has started"),
+		EventType: api.String("Datadog Operator Startup"),
+	}
+	if _, err := dd.datadogClient.PostEvent(event); err != nil {
+		return err
+	}
+	return nil
+}
+
+// initAPIClient initializes and validates the Datadog API client
+func (dd *MetricsForwarder) initAPIClient(apiKey, appKey string) error {
+	if dd.delegator == nil {
+		dd.delegator = dd
+	}
+	datadogClient, err := dd.validateCreds(apiKey, appKey)
+	if err != nil {
+		return err
+	}
+	dd.datadogClient = *datadogClient
+	dd.keysHash = hashKeys(apiKey, appKey)
+	return nil
+}
+
+// updateCredsIfNeeded used to update Datadog apiKey and appKey if they change
+func (dd *MetricsForwarder) updateCredsIfNeeded(apiKey, appKey string) error {
+	if dd.keysHash != hashKeys(apiKey, appKey) {
+		return dd.initAPIClient(apiKey, appKey)
+	}
+	return nil
+}
+
+// validateCreds returns validates the creds by querying the Datadog API
+func (dd *MetricsForwarder) validateCreds(apiKey, appKey string) (*api.Client, error) {
+	return dd.delegator.delegatedValidateCreds(apiKey, appKey)
+}
+
+// delegatedValidateCreds is separated from validateCreds to facilitate mocking the Datadog API
+func (dd *MetricsForwarder) delegatedValidateCreds(apiKey, appKey string) (*api.Client, error) {
+	datadogClient := api.NewClient(apiKey, appKey)
+	valid, err := datadogClient.Validate()
+	if !valid || err != nil {
+		return nil, fmt.Errorf("datadog client is invalid: %v", err)
+	}
+	return datadogClient, nil
+}
+
+// sendStatusMetrics forwards metrics for each component deployment (agent, clusteragent, clustercheck runner)
+// based on the status of DatadogAgentDeployment
+func (dd *MetricsForwarder) sendStatusMetrics(status *datadoghqv1alpha1.DatadogAgentDeploymentStatus) error {
+	if status == nil {
+		return errors.New("nil status")
+	}
+	var metricValue float64
+
+	// Agent deployment metrics
+	if status.Agent != nil {
+		if status.Agent.Available == status.Agent.Desired {
+			metricValue = deploymentSuccessValue
+		} else {
+			metricValue = deploymentFailureValue
+		}
+		tags := append(dd.tags, fmt.Sprintf(stateTagFormat, string(status.Agent.State)))
```

Comment:
Minor nit - since this gets repeated below, could add a helper here - e.g. `tagsWithState(..) []string`, then `tags := dd.tagsWithState(status.Agent.State)`